hoomd/md/pytest/test_custom_force.py

```diff
@@ -0,0 +1,323 @@
+        npt.assert_allclose(forces1, forces2)
+
+    energies1 = integrator.forces[0].energies
+    energies2 = integrator2.forces[0].energies
+    if sim.device.communicator.rank == 0:
+        npt.assert_allclose(energies1, energies2)
+
+    torques1 = integrator.forces[0].torques
+    torques2 = integrator2.forces[0].torques
+    if sim.device.communicator.rank == 0:
+        npt.assert_allclose(torques1, torques2)
+
+    assert integrator.forces[0].virials == integrator2.forces[0].virials
+
+
+class NestedForceCPU(md.force.Custom):
+
+    def __init__(self):
+        super().__init__()
+
+    def set_forces(self, timestep):
+        with self.cpu_local_force_arrays:
+            with self.cpu_local_force_arrays:
+                return
+
+
+class NestedForceGPU(md.force.Custom):
+
+    def __init__(self):
+        super().__init__()
+
+    def set_forces(self, timestep):
+        with self.gpu_local_force_arrays:
+            with self.gpu_local_force_arrays:
+                return
+
+
+@pytest.mark.parametrize("force_cls", [NestedForceCPU, NestedForceGPU],
+                         ids=lambda x: x.__name__)
+def test_nested_context_managers(force_cls, two_particle_snapshot_factory,
+                                 simulation_factory):
+    snap = two_particle_snapshot_factory()
+    sim = simulation_factory(snap)
+    custom_force = force_cls()
+    nvt = md.methods.NPT(hoomd.filter.All(),
+                         kT=1,
+                         tau=1,
+                         S=1,
+                         tauS=1,
+                         couple="none")
+    integrator = md.Integrator(dt=0.005, forces=[custom_force], methods=[nvt])
+    sim.operations.integrator = integrator
+    with pytest.raises(RuntimeError):
+        sim.run(1)
+
+
+class GhostForceAccessCPU(md.force.Custom):
+
+    def __init__(self):
+        super().__init__()
+        self._array_buffers = []
+        for buffer in ['force', 'torque', 'potential_energy', 'virial']:
+            self._array_buffers.append(buffer)
+            self._array_buffers.append(buffer + '_with_ghost')
+            self._array_buffers.append('ghost_' + buffer)
+
+    def _check_buffer_lengths(self, arrays):
+        """Ensure the lengths of the accessed buffers are right."""
+        for buffer_name in self._array_buffers:
+            buffer = getattr(arrays, buffer_name)
+            if buffer_name.endswith('_with_ghost'):
+                assert len(buffer) == 2
+            else:
+                assert len(buffer) == 1
+
+    def set_forces(self, timestep):
+        with self.cpu_local_force_arrays as arrays:
+            self._check_buffer_lengths(arrays)
+
+
+class GhostForceAccessGPU(GhostForceAccessCPU):
+
+    def __init__(self):
+        super().__init__()
+
+    def set_forces(self, timestep):
+        with self.gpu_local_force_arrays as arrays:
+            self._check_buffer_lengths(arrays)
+
+
+@pytest.mark.parametrize("force_cls",
```

Comment:
I could use a description here because I'm not as familiar with the internals for Ghost particles and MPI. I think it might help other users too when they at correctly accessing these values for MPI runs. I think the main point here is that we don't need to have any portion protected by rank==1, correct?

Reply:
The reason for this test is that we have to make sure when we access that `ghost_XXX` and `XXX_with_ghost` attributes of the custom force, they have the right data in them. This test uses an LJ force to make sure there is ghost data available and then just checks to make sure the lengths of the ghost buffers have the right number of particles in them.

I could add a small comment in the docstring for this test to make sure that's clearer. Teaching users how the ghost data is organized and best accessed with custom forces is best left for a tutorial.